src/vma/util/agent.cpp

```diff
@@ -227,18 +230,190 @@ agent::~agent()
 	}
 }
 
+void agent::register_cb(agent_cb_t fn, void *arg)
+{
+	agent_callback_t *cb;
+	struct list_head *entry;
+
+	if (AGENT_CLOSED == m_state) {
+		return ;
+	}
+
+	if (NULL == fn) {
+		return ;
+	}
+
+	m_cb_lock.lock();
+	/* check if it exists in the queue */
+	list_for_each(entry, &m_cb_queue) {
+		cb = list_entry(entry, agent_callback_t, item);
+		if ((cb->cb == fn) && (cb->arg == arg)) {
+			m_cb_lock.unlock();
+			return ;
+		}
+	}
+	/* allocate new callback element and add to the queue */
+	cb = (agent_callback_t *)calloc(1, sizeof(*cb));
+	if (cb) {
+		cb->cb = fn;
+		cb->arg = arg;
+		list_add_tail(&cb->item, &m_cb_queue);
+	}
+	m_cb_lock.unlock();
+}
+
+void agent::unregister_cb(agent_cb_t fn, void *arg)
+{
+	agent_callback_t *cb;
+	struct list_head *entry;
+
+	if (AGENT_CLOSED == m_state) {
+		return ;
+	}
+
+	m_cb_lock.lock();
+	/* find element in the queue and remove one */
+	list_for_each(entry, &m_cb_queue) {
+		cb = list_entry(entry, agent_callback_t, item);
+		if ((cb->cb == fn) && (cb->arg == arg)) {
+			list_del_init(&cb->item);
+			free(cb);
+			m_cb_lock.unlock();
+			return ;
+		}
+	}
+	m_cb_lock.unlock();
+}
+
+int agent::put(const void *data, size_t length, intptr_t tag, void **saveptr)
+{
+	agent_msg_t *msg = NULL;
+	int i = 0;
+
+	if (AGENT_CLOSED == m_state) {
+		return 0;
+	}
+
+	if (m_sock_fd < 0) {
+		return -EBADF;
+	}
+
+	if (length > sizeof(msg->data)) {
+		return -EINVAL;
+	}
+
+	m_msg_lock.lock();
+
+	if (saveptr) {
+		msg = (agent_msg_t *)(*saveptr);
+	}
+
+	if ((AGENT_ACTIVE == m_state) ||
+			((saveptr) &&
```

Comment:
`saveptr` is a part of current agent::put() api. It is not used for this specifc case but it might be used for future.